### PR TITLE
Fixed #32704 -- Fixed list of deferred fields when chaining QuerySet.defer() after only().

### PR DESCRIPTION
**Pull request by David-Wobrock**
Hi 👋 

Ticket https://code.djangoproject.com/ticket/32704

Not all added test assertions are entirely relevant to the patch, but it felt like they were missing and are completing the test method.
Heavily inspired by Mariusz' patch https://code.djangoproject.com/attachment/ticket/32704/ticket-32704.diff

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -2090,7 +2090,14 @@ def add_deferred_loading(self, field_names):
             self.deferred_loading = existing.union(field_names), True
         else:
             # Remove names from the set of any existing "immediate load" names.
-            self.deferred_loading = existing.difference(field_names), False
+            new_existing = existing.difference(field_names)
+            if not new_existing:
```
Comment (cjerdonek) [inline django/db/models/sql/query.py:2094]: If you handle the positive case first, this can be the simpler `if new_existing:`. You could also use an assignment expression to simplify further (`if new_existing := existing.difference(field_names):`).

File: tests/defer/tests.py [hunk 2]
```diff
@@ -49,8 +49,15 @@ def test_defer_only_chaining(self):
         qs = Primary.objects.all()
         self.assert_delayed(qs.only("name", "value").defer("name")[0], 2)
         self.assert_delayed(qs.defer("name").only("value", "name")[0], 2)
+        self.assert_delayed(qs.defer("name").only("name").only("value")[0], 2)
```
Comment (felixxm) [inline tests/defer/tests.py:52]: This works without the patch and should be moved to a separate commit.

File: tests/defer/tests.py [hunk 3]
```diff
@@ -49,8 +49,15 @@ def test_defer_only_chaining(self):
...
         self.assert_delayed(qs.defer("name").only("value")[0], 2)
         self.assert_delayed(qs.only("name").defer("value")[0], 2)
+        self.assert_delayed(qs.only("name").defer("name").defer("value")[0], 1)
+        self.assert_delayed(qs.only("name").defer("name", "value")[0], 1)
+
+        # Cancel each other out
+        self.assert_delayed(qs.only("name").defer("name")[0], 0)
+        self.assert_delayed(qs.defer("name").only("name")[0], 0)
```
Comment (felixxm) [inline tests/defer/tests.py:60]: This works without the patch and should be moved to a separate commit. I'd also move it to a separate test method:
```python
def test_defer_only_clear(self):
    self.assert_delayed(qs.only('name').defer('name')[0], 0)
    self.assert_delayed(qs.defer('name').only('name')[0], 0)
```

File: django/db/models/sql/query.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/sql/query.py:2094]: Agreed, you can use the walrus operator even twice:
```python
            if new_existing := existing.difference(field_names):
                self.deferred_loading = new_existing, False
            else:
                self.clear_deferred_loading()
                if new_only := set(field_names).difference(existing):
                    self.deferred_loading = new_only, True
```

**Review by felixxm — COMMENTED** [review]
@David-Wobrock Thanks for this patch :+1:

**Comment by David-Wobrock** [comment]
Hi @cjerdonek , hi @felixxm 

Thanks a lot for the reviews :muscle: I took them into account.
Smart use of the walrus operator, I wasn't aware that Django would drop Python 3.7 in next releases.

I suspect that the commit messages are not adequate w.r.t. Django's convention - I still don't get the hang of it :grimacing: Please feel free renaming them :pray:

**Comment by felixxm** [comment]
@David-Wobrock Thanks for updates :+1:
